### PR TITLE
[Orderbook] transitive closure

### PR DESCRIPTION
**Pull request by fleupold**
This PR finally adds the logic to compute the transitive closure of two "connecting" orderbooks (e.g. ETH/DAI & DAI/USDC => ETH/USDC).

I wrote the logic to combine all asks of two orderbooks. I then used inversion and addition to turn the bids into asks and reuse the same logic.

I also added a unit test with an extensive description on what is going on. For me, this was quite hard to grasp cognitively. I'm therefore worried that the review might be hard.

Please take some time to think this through and be generous with suggestions on how to make this easier to understand for future maintainers.

🙏 

In order to avoid rounding errors from multiple inversions I had to make the volume a Fraction (and ended up implementing a bunch of methods on this datatype). If it's easier I can factor this into its own PR.

### Test Plan

Unit test

### CODE REVIEW

File: src/orderbook.ts [hunk 1]
```diff
@@ -99,18 +139,93 @@ export class Orderbook {
...
+        right_offer.volume = right_offer.volume.sub(
+          volume.mul(left_offer.price)
+        );
+        left_next = left_iterator.next();
+        // In case the orders matched perfectly we will move right as well
+        if (right_offer.volume.toNumber() == 0) {
+          right_next = right_iterator.next();
+        }
+      }
+      result.addAsk(new Offer(price, volume));
```
Comment (nlordell) [inline src/orderbook.ts:152]: Do fees need to be considered in these calculations somehow? The reason I say that is because as you create transitive closures, the actual volume available gets smaller and smaller.

**Review by nlordell — APPROVED** [review]
Looks great! Just have one question -

File: src/orderbook.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import {Order} from ".";
 
-export class Price {
+export class Fraction {
```
Comment (nlordell) [inline src/orderbook.ts]: Maybe this can be moved into its own file so all the rational number math is outside of the oderbook code.

File: src/orderbook.ts [hunk 3]
```diff
@@ -9,8 +9,44 @@ export class Price {
     this.denominator = denominator;
   }
 
+  shorten() {
```
Comment (nlordell) [inline src/orderbook.ts]: naming nit: `reduced`.

File: src/orderbook.ts (hunk 2 above)
Comment (nlordell) [inline src/orderbook.ts]: If this fraction is being used to represent rational numbers, should this be using `BN` instead of plain `number`? This would help avoid rounding errors for large numerators and denominators.

File: src/orderbook.ts [hunk 4]
```diff
@@ -99,18 +139,93 @@ export class Orderbook {
...
+
+    ask_closure.add(bid_closure);
+    return ask_closure;
+  }
+
+  private transitiveAskClosure(orderbook: Orderbook) {
+    const result = new Orderbook(this.baseToken, orderbook.quoteToken);
+
+    // Create a copy here so original orders stay untouched
+    const left_asks = Array.from(this.asks.values());
```
Comment (nlordell) [inline src/orderbook.ts:118]: This might not be a deep clone, in that modifying the fractions might modify it in both arrays. I'm not sure though - my JS knowledge is not super great.

File: src/orderbook.ts [hunk 5]
```diff
@@ -99,18 +139,93 @@ export class Orderbook {
...
+    let right_next = right_iterator.next();
+    let left_next = left_iterator.next();
+    while (!(left_next.done || right_next.done)) {
+      const right_offer = right_next.value;
+      const left_offer = left_next.value;
+      const price = left_offer.price.mul(right_offer.price);
+      let volume;
+      const right_offer_volume_in_left_offer_base_token = right_offer.volume.div(
+        left_offer.price
+      );
```
Comment (nlordell) [inline src/orderbook.ts:136]: ~~Love the `;` but @bh2smith might have a heart-attack. Not sure what `tsc` has to say though.~~

**edit**: should be the subject of a separate PR where linting and code formatting for TS get added.

File: src/orderbook.ts (hunk 2 above)
Comment (nlordell) [inline src/orderbook.ts]: nit: `Fraction` contructor should throw if `denominator` is 0.

File: src/orderbook.ts [hunk 6]
```diff
@@ -99,18 +139,93 @@ export class Orderbook {
...
+    const ask_closure = this.transitiveAskClosure(orderbook);
+
+    // Since bids are the asks of the inverted orderbook, computing transitive closure of bids is equivalent to
+    // 1) inverting both orderbooks
+    // 2) computing the transitive ask closure on the inverses
+    // 3) re-inverting the result
+    const bid_closure = orderbook
+      .inverted()
+      .transitiveAskClosure(this.inverted())
+      .inverted();
```
Comment (nlordell) [inline src/orderbook.ts:108]: This is :sunglasses:

File: src/orderbook.ts [hunk 7]
```diff
@@ -122,11 +237,25 @@ function invertPricePoints(prices: Map<number, Offer>) {
...
+  if (num1 === 0 || num2 === 0) {
+    return 0;
+  }
+  if (num1 === num2) {
+    return num1;
+  }
+  if (num1 > num2) {
+    return gcd(num1 - num2, num2);
+  }
+  return gcd(num1, num2 - num1);
```
Comment (nlordell) [inline src/orderbook.ts]: :crossed_fingers: V8 has TCO

File: test/models/orderbook.spec.ts [hunk 8]
```diff
@@ -126,4 +126,58 @@ describe("Orderbook", () => {
...
+
+    assert.equal(closure.pair(), "ETH/USDC");
+    assert.equal(
+      JSON.stringify(closure.toJSON()),
+      JSON.stringify({
+        // The best bid eth_dai has enough liquidity on the best dai_usdc bid
+        // with 1 DAI remaining to be matched from the second best eth_dai bid (190 DAI).
+        // The remainder of the second best eth_dai bid (189 DAI) gets matched with the second
+        // best dai_usdc bid (200 DAI), leaving 11 DAI for the worst eth_dai bid (270 DAI).
+        // The remaining 259 DAI are unmatchable.
```
Comment (nlordell) [inline test/models/orderbook.spec.ts:156]: Transitive orderbooks are complicated, but I think this comment does a good job at explaining why the expected values are what they are. :1st_place_medal:

**Review by nlordell — APPROVED** [review]

**Comment by fedgiac** [comment]
Very nicely written, I always had the feeling I was knowing what was going to happen next. I also believe the logic to be solid.

My main concern is about rounding errors. I see that the test uses very low values in the fractions, but if the price is a fraction with numerator and denominator in the order of 10^18 then sometimes `gcd` can fail. For example:
```
n1 = 10**18
n2 = 10**18+10**16
m1 = 10**18
m2 = 10**18
a = new Fraction(n1, n2)
b = new Fraction(m1, m2)
a.sub(b) // gcd fails
```
There are also some assumptions like: if frac1.toNumber() > frac2.toNumber() then frac1.sub(frac2) > 0, which I'm not really so sure about when considering rounding.
Even if the numbers we expect are low, it's easy for them to grow when the denominator is the product of many denominators, like when chaining a transitivized orderbook with another one.

I assumed that the orderbook prices already includes the exchange fee. In this case the logic should be good.

File: src/orderbook.ts (hunk 6 above)
Comment (fedgiac) [reply src/orderbook.ts:108]: Very nice!

**Review by fedgiac — APPROVED** [review]

**Comment by nlordell** [comment]
> I assumed that the orderbook prices already includes the exchange fee. In this case the logic should be good.

I think you might be right. I wasn't sure when I asked, just wanted to make sure that it was considered.

File: src/orderbook.ts (hunk 4 above)
Comment (fleupold) [reply src/orderbook.ts:118]: You are right that this is not a deep clone.

All the methods on fraction return new instances and by making numerator/denominator private they can not be modified from outside.

However, you are right that this is not enforced by the compiler for the future. The way to do this would be to implement a clone method on all our datatypes and invoke this for each Offer. Not sure this is worth it.

File: src/orderbook.ts (hunk 5 above)
Comment (fleupold) [reply src/orderbook.ts:136]: Agree. We need to change some linter setting. Right now when I omit the `;` it gets auto-added on save.

File: src/orderbook.ts (hunk 7 above)
Comment (fleupold) [reply src/orderbook.ts]: Doesn't seem so: https://bugs.chromium.org/p/v8/issues/detail?id=4698

Yet, I don't think this will be our bottleneck when at the same time we are querying data from the blockchain

File: src/orderbook.ts (hunk 4 above)
Comment (nlordell) [reply src/orderbook.ts:118]: You are right, I don't think its worth it. You should mark the Fraction properties as `readonly`. That way they can only be assigned in the constructor and the compiler guarantees they are not mutated.

File: src/orderbook.ts (hunk 7 above)
Comment (nlordell) [reply src/orderbook.ts]: It really shouldn't matter at all. I studied embedded programming so when I see recursion a little alarm goes off in my brain.

File: src/orderbook.ts (hunk 2 above)
Comment (nlordell) [inline src/orderbook.ts]: Also, I just remembered, TS has this cool feature:
```ts
export class Fraction {
    constructor(
        public readonly numerator: number,
        public readonly denominator: number,
    ) {
        if (denominator === 0) {
            throw new Error("error");
        }
    }
}
```

This will make both the properties read-only and automagically set their value from the constructor arguments.

Comment (nlordell) [inline src/orderbook.ts]: Also, I just remembered, TS has this cool feature:
```ts
export class Fraction {
    constructor(
        public readonly numerator: number,
        public readonly denominator: number,
    ) {
        if (denominator === 0) {
            throw new Error("error");
        }
    }
}
```

This will make both the properties read-only and automagically set their value from the constructor arguments.

File: src/orderbook.ts [hunk 9]
```diff
@@ -9,8 +9,44 @@ export class Price {
     this.denominator = denominator;
   }
```
Comment (nlordell) [inline src/orderbook.ts]: Also, I just remembered, TS has this cool feature:
```ts
export class Fraction {
    constructor(
        public readonly numerator: number,
        public readonly denominator: number,
    ) {
        if (denominator === 0) {
            throw new Error("error");
        }
    }
}
```

This will make both the properties read-only and automagically set their value from the constructor arguments.

File: src/orderbook.ts (hunk 1 above)
Comment (fleupold) [reply src/orderbook.ts:152]: Yes that is true. I left it our here initially as wrapping my head around the logic in the absence of a fee seemed simpler and already hard enough. I'm planning to implement this e2e without a fee for now but the later add a fee as a construction argument to the orderbook (or as a global param). We can then reduce the available amount when we multiply the asks on each round.

File: src/orderbook.ts (hunk 4 above)
Comment (fleupold) [reply src/orderbook.ts:118]: `reduce` reassigns the numerator and denominator by dividing them with the gcd, thus readonly doesn't work. Moreover, since BN also has some in-place methods (e.g. imul, iadd, but ironically not idiv at least in the typescript bindings) the values can be mutated even if they are not reassigned.

Rust is a lot stronger when it comes to type/mutation safety.

File: src/orderbook.ts (hunk 9 above)
Comment (fleupold) [reply src/orderbook.ts]: This is cool. I wanted to allow instantiation with number as well as with BN, so I had to create some matching logic in the constructor.

File: src/fraction.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
+import BN from "bn.js";
+
+export class Fraction {
+  private numerator: BN;
+  private denominator: BN;
+
+  constructor(numerator: BN | number, denominator: BN | number) {
+    if (typeof numerator == "number") {
+      this.numerator = new BN(numerator);
```
Comment (nlordell) [inline src/fraction.ts]: I think this if statement is not necessary: 
```
> new BN(5)
<BN: 5>
> new BN(new BN(5))
<BN: 5>
```

In the second format, there is no allocation, so:
```
> a = new BN(1)
<BN: 1>
> b = new BN(2)
<BN: 2>
> c = new BN(a)
<BN: 1>
> c.iadd(b)
<BN: 3>
> a
<BN: 3>
```

File: src/fraction.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+      this.denominator = denominator;
+    }
+
+    if (this.denominator.isZero()) {
+      throw Error("Denominator cannot be zero");
+    }
+  }
+
+  reduce() {
+    const greatest_common_denominator = gcd(this.numerator, this.denominator);
```
Comment (nlordell) [inline src/fraction.ts]: nit: `this.numerator.gcd(this.denominator)` - BN.js has an implementation.

File: src/fraction.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+  sub(other: Fraction) {
+    return this.add(other.negated());
+  }
+
+  add(other: Fraction) {
+    const result = new Fraction(
+      this.numerator
+        .mul(other.denominator)
+        .add(other.numerator.mul(this.denominator)),
```
Comment (nlordell) [inline src/fraction.ts]: uber-tiny-mini-nit:
```suggestion
        .iadd(other.numerator.mul(this.denominator)),
```
Modifies the result of `mul` in-place.

File: src/fraction.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
+import BN from "bn.js";
+
+export class Fraction {
+  private numerator: BN;
+  private denominator: BN;
```
Comment (nlordell) [inline src/fraction.ts:5]: Maybe these can be `readonly` instead of private?

**Review by nlordell — APPROVED** [review]

**Comment by fleupold** [comment]
@fedgiac thanks for the soundness analysis, you are right that using small examples for the tests might oversee some large number implications.

I changed Fraction to now be based on BigNumber and instead of using `frac1.toNumber() > frac2.toNumber()` I implemented a proper `gt` method on Fraction. Does this lift your concerns or am I overseeing something?

I will also add some unittest for the Fraction class in a follow up PR.

File: src/fraction.ts (hunk 13 above)
Comment (fleupold) [reply src/fraction.ts:5]: The problem is the `reduce` method, where I reassign the numerator/denominator. Do you think I should return a new Fraction there as well?

Comment (nlordell) [reply src/fraction.ts:5]: Hmmm, didn't think of that. No, I think you are right.

File: src/orderbook.ts (hunk 4 above)
Comment (nlordell) [reply src/orderbook.ts:118]: It has `idivn` though :stuck_out_tongue: 

And really great points, Its nice the way it is!

File: src/fraction.ts [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    this.numerator = new BN(numerator);
+    this.denominator = new BN(denominator);
+
+    if (this.denominator.isZero()) {
+      throw Error("Denominator cannot be zero");
+    }
+  }
+
+  gt(other: Fraction) {
+    return this.sub(other).toNumber() > 0;
```
Comment (fedgiac) [inline src/fraction.ts]: I would rather avoid relying on floats since we are using BN anyway. Another way to rephrase this comparison using only `BN` could be:
```
return this.numerator.mul(other.denominator).gt(other.numerator.mul(this.denominator))
```

File: src/orderbook.ts [hunk 15]
```diff
@@ -230,7 +170,7 @@ function addOffer(offer: Offer, existingOffers: Map<number, Offer>) {
 }
 
 function sortOffersAscending(left: Offer, right: Offer) {
-  return left.price.toNumber() - right.price.toNumber();
+  return left.price.sub(right.price).toNumber();
```
Comment (fedgiac) [inline src/orderbook.ts]: Maybe this should use Fraction comparisons? If you implement my proposed changes to `.gt` then also this should change.

File: src/orderbook.ts [hunk 16]
```diff
@@ -99,30 +82,102 @@ export class Orderbook {
...
+        left_offer.volume = left_offer.volume.sub(volume);
+        right_next = right_iterator.next();
+      } else {
+        volume = left_offer.volume;
+        right_offer.volume = right_offer.volume.sub(
+          volume.mul(left_offer.price)
+        );
+        left_next = left_iterator.next();
+        // In case the orders matched perfectly we will move right as well
+        if (right_offer.volume.toNumber() == 0) {
```
Comment (fedgiac) [inline src/orderbook.ts]: You could also do this check as `right_offer.volume.numerator.isZero()`.

File: src/fraction.ts (hunk 14 above)
Comment (fedgiac) [reply src/fraction.ts]: Sorry, this is actually a bad recommendation since the denominators could be negative.
A working comparison is:
```
const diff = this.sub(other)
return diff.numerator.mul(diff.denominator).gt(new BN(0))
```
You multiply numerator and denominator of the difference. If the sign is positive, then the difference is positive, otherwise it's not.

File: src/orderbook.ts (hunk 15 above)
Comment (fleupold) [reply src/orderbook.ts]: The sorting function needs to return a number:

> It is expected to return a negative value if first argument is less than second argument, zero if they're equal and a positive value otherwise.

I can also implement `lt` on Fraction and use a combination of `gt` and `lt`.

Comment (fedgiac) [reply src/orderbook.ts]: I'd do it just for the peace of mind of not worrying about consistency between sorting and `.gt`. On the other end, for what I understand about float division you could have an inconsistent result only when your fractions have a numerator/denominator ratio of more than ~10^162, so maybe we don't want to worry about that.

Comment (fleupold) [reply src/orderbook.ts]: It's done! Thanks for the suggestions.

File: src/fraction.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
+import BN from "bn.js";
+
+export class Fraction {
```
Comment (bh2smith) [inline src/fraction.ts:3]: Do classes like this not already exist in a standard library somewhere? Was there something about the implementation here that is so different from others?

Comment (fleupold) [reply src/fraction.ts:3]: It has to work with BN as numerator/denominator
